tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/localedistance/DistanceTable.java

```diff
@@ -0,0 +1,556 @@
+            } else if (key.isWildcard()) {
+                // An intermediate wildcard mapping is applied to all existing sub-nodes.
+                // NOTE: This will need to change if we want to support "mixed" wildcard mappings.
+                for (Node node : subtables.values()) {
+                    node.addWildcardMappings(key.getSuffix());
+                }
+            } else {
+                // An explicit intermediate mapping only affects an existing exact match.
+                Node node = getNode(key);
+                if (node != null) {
+                    node.addWildcardMappings(key.getSuffix());
+                }
+            }
+        }
+
+        /**
+         * Adds a new mapping to this node with the specified distance if it didn't already
+         * exist.
+         *
+         * <p>Note: If a mapping already exists, then this method has no effect (even if the
+         * existing distance differs from the given distance). This is necessary to for two
+         * reasons:
+         * <ol>
+         *     <li>An earlier match rule may have set an explicit value for the mapping,
+         *         and we subsequently try to set a default value (via a wildcard mapping).
+         *         This should be ignored, since we want the non-default value to win.
+         *         This means it's important to always have explicit {@code <languageMatch>}
+         *         rules before any related wildcard rules in the CLDR data.
+         *
+         *     <li>A preferential {@code <languageMatch>} rule appears earlier in CLDR data.
+         *         This occurs because of the way partitions are defined and allows for two
+         *         distinct {@code <languageMatch>} rules to generate the same mapping (with
+         *         different distances). This is because region variables reference sets of
+         *         partition IDs and these are not always disjoint (e.g. "en_*_$!enUS" and
+         *         "en_*_GB" both contain the partition ID for "GB").
+         * </ol>
+         *
+         * @return true if a new mapping was added, or if the distances were equal (i.e.
+         *         the operation was idempotent).
+         */
+        private boolean putIfAbsent(MappingKey key) {
+            Node node = getNode(key);
+            if (node == null) {
+                logger.fine(() -> String.format("add: %s", key));
+                subtables.put(key.getDesired(), key.getSupported(), new Node(key.getDistance()));
+                return true;
+            }
+            return (key.getDistance() == node.distance);
+        }
+
+        /**
+         * Returns a sub-node corresponding to the given {@code (desired, supported)} mapping.
+         * If the node already exists, it is simply returned, otherwise a new node is created
+         * and any existing wildcard mappings are copied into it.
+         */
+        private Node getIntermediateNode(MappingKey key) {
```

Comment:
This code (and its caller "addExplicitMapping()") are where some of the biggest reductions in complexity happen in this PR.

Main original logic (until about line 350). Using the MappingKey avoids having to duplicate almost identical logic in several places (for different combinations of language/script/region) since now every level in the table is conceptually the same (just with a different key).
https://github.com/unicode-org/icu/blob/master/icu4j/tools/misc/src/com/ibm/icu/dev/tool/locale/LocaleDistanceBuilder.java#L202

Esoteric classes for callbacks to copy default data (which are no longer needed at all by virtue of restructuring the code a bit).
https://github.com/unicode-org/icu/blob/master/icu4j/tools/misc/src/com/ibm/icu/dev/tool/locale/LocaleDistanceBuilder.java#L434

One of the notable simplifications is the removal of the unreachable code related to fallback checking. Since there were previous checks to ensure that (desired==ANY if-and-only-if supported=ANY), it was never possible to find a fallback match of <X,ANY> or <ANY,X> in the table. This means that the code in "getDistance()" at line 202 can be replaced by 3 lines of code here (getNode(key), check-null, getAnyNode()).

The remaining parts of this method come from the original callers of "getDistance()", which used the special "CopyIfEmpty" and "AddSub" class closures to hack around deciding when to copy data into new nodes. Since the new logic means the point at which a copying should be done is obvious, none of this is needed.